lib/core/src/Ouroboros/Network/Client/Wallet.hs

```diff
@@ -475,62 +494,97 @@ type LocalStateQueryResult state = Either AcquireFailure state
 
     clientStAcquiring
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
+        :: LocalStateQueryCmd block m
+        -> Point block
+        -> AnyCardanoEra
+        -> Int
         -> LSQ.ClientStAcquiring block (Point block) (Query block) m Void
-    clientStAcquiring query respond = LSQ.ClientStAcquiring
-        { recvMsgAcquired = clientStAcquired query respond
-        , recvMsgFailure = \failure -> do
-                respond (Left failure)
-                clientStIdle
+    clientStAcquiring qry oldPt era retries = LSQ.ClientStAcquiring
+        { recvMsgAcquired = clientStAcquired qry era
+        , recvMsgFailure = \_failure -> do
+            (newEra, newPt) <- second getTipPoint <$> (atomically $ do
+                t <- readTVar tipVar
+                guard (getTipPoint (snd t) /= oldPt)
+                return t)
+            pure $ LSQ.SendMsgAcquire newPt (clientStAcquiring qry newPt newEra (retries + 1))
         }
 
     clientStAcquired
-        :: forall state. Query block state
-        -> (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStAcquired block (Point block) (Query block) m Void
-    clientStAcquired query respond =
-        LSQ.SendMsgQuery query (clientStQuerying respond)
-
-    -- By re-acquiring rather releasing the state with 'MsgRelease' it
-    -- enables optimisations on the server side.
-    clientStAcquiredAgain
-        :: m (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
-    clientStAcquiredAgain = awaitNextCmd <&> \case
-        CmdQueryLocalState pt query respond ->
-            LSQ.SendMsgReAcquire pt (clientStAcquiring query respond)
-
-    clientStQuerying
-        :: forall state. (LocalStateQueryResult state -> m ())
-        -> LSQ.ClientStQuerying block (Point block) (Query block) m Void state
-    clientStQuerying respond = LSQ.ClientStQuerying
-        { recvMsgResult = \result -> do
-            respond (Right result)
-            clientStAcquiredAgain
-        }
+        :: LocalStateQueryCmd block m
+        -> AnyCardanoEra
+        -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+    clientStAcquired (SomeLSQ cmd respond) era = go cmd $ \res -> do
+        LSQ.SendMsgRelease (respond res >> clientStIdle)
+        -- TODO: We might not want to release. Re-acquiring, or even staying
+        -- acquired should be more efficient better.
+      where
+          go
+              :: forall a. LSQ block m a
+              -> (a -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void))
+              -> (LSQ.ClientStAcquired block (Point block) (Query block) m Void)
+          go (LSQPure a) cont = (cont a)
+          go (LSQry qry) cont = LSQ.SendMsgQuery qry $ LSQ.ClientStQuerying $ \res -> do
+              pure $ cont res
+          go (LSQBind ma f) cont = go ma $ \a -> do
+              go (f a) $ \b -> cont b
+          go (LSQEra) cont = cont era
```

Comment:
Hmmm. The continuation above will `respond`, that is, push a message into the response queue. This means that it is possible for callers to create "ill-programs" that will generate many responses in the queue. For instance `LsqPure a >> LsqPure a` will yield two messages. That's pretty bad as we may both create a memory leak but also fill the queue with responses which may no longer correspond to queries. 

Unless I am missing something?

Reply:
Ah, now I get what you mean. No — I don't think so.

The final `respond` will only be called by the top-level `go`.

```
LSQPure a >> LSQPure a
== LSQBind (LSQPure a) (\_ -> LSQPure a)
```

which should evaluate something like this:
```
go ma $ \a -> go (f a) $ \b -> cont b
go (LSQPure a) $ \a -> go (f a) $ \b -> cont b
go (f a) $ \b -> cont b
go (LSQPure a) $ \b -> cont b
cont a
respond a
```

Reply:
Could perhaps abstract the run-function to the top level and test it — but not sure how, on the top off my head.

Reply:
Also: The danger I see is if the queries contain a partial function. If it throws, I think the entire state machine will come to a halt.

Reply:
Luckily the entire wallet crashes in that case:
```
[cardano-wallet.network:Notice:282] [2021-01-18 19:34:23.67 UTC] Query stakePoolsSummary took 19.62426s (too slow)
cardano-wallet: ExceptionInLinkedThread ThreadId 12 todo
CallStack (from HasCallStack):
  error, called at src/Cardano/Wallet/Shelley/Network.hs:502:18 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley.Network
  withNetworkLayerBase, called at src/Cardano/Wallet/Shelley/Network.hs:294:5 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley.Network
  withNetworkLayer, called at src/Cardano/Wallet/Shelley.hs:274:9 in cardano-wallet-2021.1.12-3Kq63VUOZn92WgCVghwpjO:Cardano.Wallet.Shelley
```